pkg/cmd/infra/deployer/deployer.go

```diff
@@ -196,5 +203,53 @@ func (d *Deployer) Deploy(namespace, deploymentName string) error {
 	} else {
 		glog.Infof("Deploying from %s to %s (replicas: %d)", deployutil.LabelForDeployment(from), deployutil.LabelForDeployment(to), desiredReplicas)
 	}
-	return strategy.Deploy(from, to, desiredReplicas)
+	if err := strategy.Deploy(from, to, desiredReplicas); err != nil {
+		return err
+	}
+	return d.after(config, to)
+}
+
+// AfterDeployment is the default post deployment behavior
+type AfterDeployment struct {
+	// tags allows setting image stream tags
+	tags client.ImageStreamTagsNamespacer
+}
+
+// Handle processes the post-deployment phase
+func (a *AfterDeployment) Handle(config *deployapi.DeploymentConfig, to *kapi.ReplicationController) error {
+	after := config.Spec.Strategy.After
+	if after == nil {
+		return nil
+	}
+
+	var errs []error
+	if len(after.TagImages) > 0 {
+		glog.Infof("Tagging images for successful deployment")
+		for _, action := range after.TagImages {
+			image, err := deployapi.TemplateImageForContainer(&to.Spec.Template.Spec, deployapi.DeploymentConfigHasTrigger(config), action.ContainerName)
```

Comment:
An admission controller that updated the RC to fully qualify pod references
(Or pods themselves) would be broken here.  I need to look at the state of
the deployment as it runs through.

On Mon, Feb 1, 2016 at 4:20 PM, Michail Kargakis <notifications@github.com>
wrote:

> In pkg/cmd/infra/deployer/deployer.go
> <https://github.com/openshift/origin/pull/6925#discussion_r51482363>:
>
> > +	// tags allows setting image stream tags
> > +	tags client.ImageStreamTagsNamespacer
> > +}
> > +
> > +// Handle processes the post-deployment phase
> > +func (a *AfterDeployment) Handle(config *deployapi.DeploymentConfig, to *kapi.ReplicationController) error {
> > +	after := config.Spec.Strategy.After
> > +	if after == nil {
> > +		return nil
> > +	}
> > +
> > +	var errs []error
> > +	if len(after.TagImages) > 0 {
> > +		glog.Infof("Tagging images for successful deployment")
> > +		for _, action := range after.TagImages {
> > +			image, err := deployapi.TemplateImageForContainer(&to.Spec.Template.Spec, deployapi.DeploymentConfigHasTrigger(config), action.ContainerName)
>
> I don't know of any other place where we touch rcs for dcs apart from the
> controllers. And there shouldn't be.
>
>    - main dc controller
>    <https://github.com/openshift/origin/blob/920ad9ab0697b9026152e5d358364533bd22eed1/pkg/deploy/controller/deploymentconfig/controller.go#L124>
>    - configchange controller
>    <https://github.com/openshift/origin/blob/920ad9ab0697b9026152e5d358364533bd22eed1/pkg/deploy/controller/configchange/controller.go#L79>
>
> Both use MakeDeployment which does the encoding of the latest dc into the
> rc.
>
>    - the deployer pod and deployment controllers update with status
>    annotations
>    - the image change controller doesn't even touch rcs
>
> That being said, I perceive the DC as the state that the latest RC should
> follow and we should guarantee that (like what happens upstream). Right now
> we don't have any strong guarantees but I think it is the case.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/6925/files#r51482363>.
>